core/src/main/scala/kafka/coordinator/group/GroupMetadata.scala

```diff
@@ -434,18 +441,51 @@ private[group] class GroupMetadata(val groupId: String, initialState: GroupState
     }.toMap
   }
 
-  def removeExpiredOffsets(startMs: Long) : Map[TopicPartition, OffsetAndMetadata] = {
-    val expiredOffsets = offsets
-      .filter {
-        case (topicPartition, commitRecordMetadataAndOffset) =>
-          commitRecordMetadataAndOffset.offsetAndMetadata.expireTimestamp < startMs && !pendingOffsetCommits.contains(topicPartition)
-      }
-      .map {
-        case (topicPartition, commitRecordOffsetAndMetadata) =>
-          (topicPartition, commitRecordOffsetAndMetadata.offsetAndMetadata)
-      }
+  def removeExpiredOffsets(currentTimestamp: Long, offsetRetentionMs: Long) : Map[TopicPartition, OffsetAndMetadata] = {
+    val expiredOffsets: Map[TopicPartition, CommitRecordMetadataAndOffset] = protocolType match {
+      case Some(protocol) if is(Empty) =>
+        // no consumer exists in the group =>
+        // if retention period has passed since group became Empty, expire all offsets with no pending offset commit
+        offsets.filter {
+          case (topicPartition, commitRecordMetadataAndOffset) =>
+            !pendingOffsetCommits.contains(topicPartition) && {
```

Comment:
I'm puzzling a little over why we only have this check in the Empty case. Is there a reason? 

Given how similar the logic between these two branches is, I wonder if we can factor out the common structure and use a function argument for the slight behavioral difference.